### ISSUE TITLE
Alembic autogeneration fails when instance_config table is changed.

### ISSUE THREAD

**Issue by zenmonkeykstop** [freedomofpress/securedrop#6312]
## Description

SecureDrop uses alembic to generate and apply database versions. Recently a partial index was added on the instance_config table to ensure there was only one row with a `valid_until` field with a null value. (SQL for this index looks like `CREATE UNIQUE INDEX ix_one_active_instance_config ON instance_config (valid_until IS NULL) WHERE valid_until IS NULL;`)

After the addition of the index, any changes to the `instance_config` table will result in a warning when autogenerating a new alembic revision via `alembic revision --autogenerate "descHere"` like:
```
...
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/sqlalchemy/dialects/sqlite/base.py:1957: SAWarning: Skipped unsupported reflection of expression-based index ix_one_active_instance_config
  util.warn(
INFO  [alembic.autogenerate.compare] Detected added column 'instance_config.initial_message_min_len'
INFO  [alembic.autogenerate.compare] Detected added column 'instance_config.reject_message_with_codename'
INFO  [alembic.autogenerate.compare] Detected added index 'ix_one_active_instance_config' on '[None]'
  Generating /home/user/securedrop/securedrop/alembic/versions/813e92344ec6_testing.py ... done
```
(Note the "unsupported reflection" warning.)

A version file will still be created, including a batch command to create the index like: 
```
 batch_op.create_index('ix_one_active_instance_config', [sa.text('valid_until IS NULL')], unique=True, sqlite_where=sa.text('valid_until IS NULL'))
```
When `alembic upgrade head` is run, it fails as follows:
```
user@73471b01877d:/home/user/securedrop/securedrop$ alembic upgrade head
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 2e24fc7536e8 -> 813e92344ec6, testing
/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/sqlalchemy/dialects/sqlite/base.py:1957: SAWarning: Skipped unsupported reflection of expression-based index ix_one_active_instance_config
  util.warn(
Traceback (most recent call last):
  File "/opt/venvs/securedrop-app-code/bin/alembic", line 8, in <module>
    sys.exit(main())
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/config.py", line 486, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/config.py", line 480, in main
    self.run_cmd(cfg, options)
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/config.py", line 461, in run_cmd
    fn(config,
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/command.py", line 254, in upgrade
    script.run_env()
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/script/base.py", line 427, in run_env
    util.load_python_file(self.dir, 'env.py')
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/util/pyfiles.py", line 81, in load_python_file
    module = load_module_py(module_id, path)
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/util/compat.py", line 83, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/user/securedrop/securedrop/alembic/env.py", line 83, in <module>
    run_migrations_online()
  File "/home/user/securedrop/securedrop/alembic/env.py", line 77, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/runtime/environment.py", line 836, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/runtime/migration.py", line 330, in run_migrations
    step.migration_fn(**kw)
  File "/home/user/securedrop/securedrop/alembic/versions/813e92344ec6_testing.py", line 24, in upgrade
    batch_op.create_index('ix_one_active_instance_config', [sa.text('valid_until IS NULL')], unique=True, sqlite_where=sa.text('valid_until IS NULL'))
  File "/usr/lib/python3.8/contextlib.py", line 120, in __exit__
    next(self.gen)
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/operations/base.py", line 300, in batch_alter_table
    impl.flush()
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/operations/batch.py", line 82, in flush
    batch_impl._create(self.impl)
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/operations/batch.py", line 277, in _create
    for idx in self._gather_indexes_from_both_tables():
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/operations/batch.py", line 215, in _gather_indexes_from_both_tables
    *[self.new_table.c[col] for col in index.columns.keys()],
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/alembic/operations/batch.py", line 215, in <listcomp>
    *[self.new_table.c[col] for col in index.columns.keys()],
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/sqlalchemy/util/_collections.py", line 194, in __getitem__
    return self._data[key]
KeyError: 'valid_until IS NULL'

```

## Steps to Reproduce

Check out `develop`, add a field to the `instance_config` table and follow the database migration process.

## Expected Behavior
`alembic upgrade head` completes successfully

## Actual Behavior
Failure as described above.

## Comments

I think there's a combo of things happening here: 
- alembic can't handle the expression-based index (as per the warning), so it doesn't drop it first (or do the whole `_tmp` move and copy thing) in batch-mode upgrades
- batch mode doesn't support expression-based indices either, and is trying to treat `valid_until IS NULL` as a basic column, hence the KeyError above. (Changing the index to just be on `valid_until` results in a passing alembic upgrade, but obviously breaks the expected index  functionality.)

Probably a workaround would involve dropping the index outside of batch mode, and recreating it outside of batch mode once the table updates are completed. But there will still be warnings on every new migration attempt, and a proper fix would remove both the warning and the need for the workaround.

(Another solution could be to reopen #5973 and solve it without the use of partial indices.)

**Comment by cfm** [comment]
It looks like this is a limitation of SQLAlchemy, rather than Alembic, with both expression-based and functional indices across several database engines, including SQLite and PostgreSQL:

- sqlalchemy/alembic#523
- sqlalchemy/sqlalchemy#4431, which links to the changes responsible for the `SAWarning: Skipped unsupported reflection of expression-based index`
- sqlalchemy/alembic#676
- sqlalchemy/alembic#750

That the problem appears to lie in, and will reoccur from, SQLAlchemy itself might argue in the long run for reopening #5973.

**Comment by legoktm** [comment]
This was fixed in #6306 by dropping the partial index and instead making it non-nullable and using the unix epoch to represent the current row.
